src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs

```diff
@@ -26,67 +26,132 @@ protected override void WriteData(IBufferWriter<byte> bufferWriter, int numBytes
+                output.Clear();
 
-            Assert.True(transientSpan.SequenceEqual(transientMemory.Span));
+                Assert.True(transientSpan[0] == 0);
+                Assert.True(transientMemory.Span[0] == 0);
 
-            Assert.True(transientSpan[0] != 0);
+                Assert.Equal(0, output.WrittenCount);
+                byte[] streamOutput = memStream.ToArray();
+
+                Assert.True(ReadOnlyMemory<byte>.Empty.Span.SequenceEqual(output.WrittenMemory.Span));
+                Assert.True(ReadOnlySpan<byte>.Empty.SequenceEqual(output.WrittenMemory.Span));
+                Assert.True(output.WrittenSpan.SequenceEqual(output.WrittenMemory.Span));
+
+                Assert.Equal(outputSpan.Length, streamOutput.Length);
+                Assert.True(outputSpan.SequenceEqual(streamOutput));
+            }
+            {
+                WriteData(output, 100);
 
-            memStream.Write(transientSpan.ToArray(), 0, transientSpan.Length);
-            output.Clear();
+                using var memStream = new MemoryStream(100);
 
-            Assert.True(transientSpan[0] == 0);
-            Assert.True(transientMemory.Span[0] == 0);
+                Assert.Equal(100, output.WrittenCount);
 
-            Assert.Equal(0, output.WrittenCount);
-            byte[] streamOutput = memStream.ToArray();
+                ReadOnlySpan<byte> outputSpan = output.WrittenMemory.ToArray();
 
-            Assert.True(ReadOnlyMemory<byte>.Empty.Span.SequenceEqual(output.WrittenMemory.Span));
-            Assert.True(ReadOnlySpan<byte>.Empty.SequenceEqual(output.WrittenMemory.Span));
-            Assert.True(output.WrittenSpan.SequenceEqual(output.WrittenMemory.Span));
+                ReadOnlyMemory<byte> transientMemory = output.WrittenMemory;
+                ReadOnlySpan<byte> transientSpan = output.WrittenSpan;
 
-            Assert.Equal(outputSpan.Length, streamOutput.Length);
-            Assert.True(outputSpan.SequenceEqual(streamOutput));
+                Assert.True(transientSpan.SequenceEqual(transientMemory.Span));
+
+                Assert.True(transientSpan[0] != 0);
+                byte t0 = transientSpan[0];
+
+                memStream.Write(transientSpan.ToArray(), 0, transientSpan.Length);
+                output.ResetWrittenCount();
+
+                Assert.True(transientSpan[0] == t0);
+                Assert.True(transientMemory.Span[0] == t0);
+
+                Assert.Equal(0, output.WrittenCount);
+                byte[] streamOutput = memStream.ToArray();
+
+                Assert.True(ReadOnlyMemory<byte>.Empty.Span.SequenceEqual(output.WrittenMemory.Span));
+                Assert.True(ReadOnlySpan<byte>.Empty.SequenceEqual(output.WrittenMemory.Span));
+                Assert.True(output.WrittenSpan.SequenceEqual(output.WrittenMemory.Span));
+
+                Assert.Equal(outputSpan.Length, streamOutput.Length);
+                Assert.True(outputSpan.SequenceEqual(streamOutput));
+            }
```

Comment:
It's a lot of copy-pasting. How about making this test a `Theory` with a boolean parameter that controls whether we use `Clear` or `ResetWrittenCount`?

```cs
[Theory]
[InlineData(true)]
[InlineData(false)]
public void WriteAndCopyToStream(bool clearContent)
{
    ArrayBufferWriter<byte> output = new();
    WriteData(output, 100);

    using MemoryStream memStream = new(100);

    Assert.Equal(100, output.WrittenCount);

    ReadOnlySpan<byte> outputSpan = output.WrittenMemory.ToArray();

    ReadOnlyMemory<byte> transientMemory = output.WrittenMemory;
    ReadOnlySpan<byte> transientSpan = output.WrittenSpan;

    Assert.True(transientSpan.SequenceEqual(transientMemory.Span));

    Assert.True(transientSpan[0] != 0);
    byte expectedFirstByte = transientSpan[0];

    memStream.Write(transientSpan.ToArray(), 0, transientSpan.Length);
    
    if (clearContent)
    {
        output.Clear();
        
        expectedFirstByte = 0;
    }
    else
    {
        output.ResetWrittenCount();
    }
    
    Assert.Equal(expectedFirstByte, transientSpan[0]);
    Assert.Equal(expectedFirstByte, transientMemory.Span[0]);
    
    Assert.Equal(0, output.WrittenCount);
    byte[] streamOutput = memStream.ToArray();

    Assert.True(ReadOnlyMemory<byte>.Empty.Span.SequenceEqual(output.WrittenMemory.Span));
    Assert.True(ReadOnlySpan<byte>.Empty.SequenceEqual(output.WrittenMemory.Span));
    Assert.True(output.WrittenSpan.SequenceEqual(output.WrittenMemory.Span));

    Assert.Equal(outputSpan.Length, streamOutput.Length);
    Assert.True(outputSpan.SequenceEqual(streamOutput));
}
```